### ISSUE TITLE
map_partitions doesn't iterate over all partitions in the case of "bad" return values of the called function

### ISSUE THREAD

**Issue by abast** [dask/dask#1518]
In the following example, map_partitions causes only 2 function calls, which is less than the number of partitions:

```
#generate example dataframe
pdf = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))
ddf = dd.from_pandas(pdf, npartitions=3)

#define helper function for map.
def helper(x):
    print('function called\n')
    return 1

#check result
out = ddf.map_partitions(helper).compute()
print(len(out))
```

I suggest, that map_partitions should raise an error in the case, that the called function does not return a pandas.Dataframe or pandas.Series object.

**Comment by mrocklin** [comment]
I believe that @jcrist tracked your StackOverflow question and resolved this in https://github.com/dask/dask/pull/1515

**Comment by jcrist** [comment]
This was indeed fixed in #1515. Instead, computations that return a non-pandas object are interpreted as a series on computation. This will cause problems if you apply other pandas methods afterwards though, as each partition is no longer a pandas object.

```python
In [1]: %load test.py

In [2]: # %load test.py
import pandas as pd
import numpy as np
import dask.dataframe as dd


pdf = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))
ddf = dd.from_pandas(pdf, npartitions=3)


def helper(x):
    print('function called\n')
    return 1


out = ddf.map_partitions(helper).compute()
print(len(out))
   ...:
function called

function called
function called


function called

3
```

Closing.
